### ISSUE TITLE
Variable is None-checked but mypy prompts `Unsupported left operand type for + ("None")`

### ISSUE THREAD

**Issue by lstolcman** [python/mypy#16799]
**Bug Report**

Consider a code checking whether all elements of a list are true. If so, it prints every element to the console with appended exclamation mark. If not, it should skip execution and print appropriate message

**To Reproduce**

example 1 - mypy returns failure

```python
li = [None, "a"]

if not all(li):
    print("Not all elems of li are true")
else:
    for elem in li:
        print(elem + "!")
```

example 2 - mypy ok

```python
li = ["a", "b"]

if not all(li):
    print("Not all elems of li are true")
else:
    for elem in li:
        print(elem + "!")
```

**Expected Behavior**
mypy should not report an error in line 7: `print(elem + "!")` - python cannot reach this code due to the condition: `if not all(li):`

**Actual Behavior**
example 1 fails:
```sh
main.py:7: error: Unsupported left operand type for + ("None")  [operator]
main.py:7: note: Left operand is of type "str | None"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: mypy 1.8.0, mypy master (from https://mypy-play.net)
- Mypy command-line flags: none, mypy-play was used
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11, 3.12
